Src/Support/Google.Apis.Auth.Tests/OAuth2/TokenRefreshManagerTests.cs

```diff
@@ -0,0 +1,283 @@
+﻿/*
+Copyright 2018 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+using Google.Apis.Auth.OAuth2;
+using Google.Apis.Auth.OAuth2.Responses;
+using Google.Apis.Logging;
+using Google.Apis.Tests.Mocks;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Google.Apis.Auth.Tests.OAuth2
+{
+    public class TokenRefreshManagerTests
+    {
+        [Theory, CombinatorialData]
+        public async Task SingleTokenConcurrentRefresh(
+            [CombinatorialValues(1, 2, 3, 6, 11)] int concurrentRefreshCount)
+        {
+            // Test multiple refreshes concurrently and sequentially,
+            // where all refreshes return the same token.
+            var clock = new MockClock { UtcNow = new DateTime(2010, 1, 1, 0, 0, 0, DateTimeKind.Utc) };
+            var logger = new NullLogger();
+            int refreshFnCount = 0;
+            TaskCompletionSource<int> delayTask = null;
+            TokenRefreshManager trm = null;
+            trm = new TokenRefreshManager(RefreshFn, clock, logger);
+            async Task<bool> RefreshFn(CancellationToken ct)
```

Comment:
Maybe move this to the end of the method, or just use a lambda expression? I was confused seeing a local method in the middle of the method.

Reply:
Alternatively, create a FakeTokenRefresher or similar, that provided access to the count and the last token issued. That might make a lot of these tests shorter.

Reply:
All changed to lambdas.